lib/identity_cache/cache_key_loader.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+module IdentityCache
+  # A generic cache key loader that supports different types of
+  # cache fetchers, each of which can use their own cache key
+  # format and have their own cache miss resolvers.
+  #
+  # Here is the interface of a cache fetcher in the
+  # [ruby-signature](https://github.com/ruby/ruby-signature)'s
+  # format.
+  #
+  # ```
+  # interface _CacheFetcher[DbKey, DbValue, CacheableValue]
+  #   def cache_key: (DbKey) -> String
+  #   def cache_encode: (DbValue) -> CacheableValue
+  #   def cache_decode: (CacheableValue) -> DbValue
+  #   def load_one_from_db: (DbKey) -> DbValue
+  #   def load_multi_from_db: (Array[DbKey]) -> Hash[DbKey, DbValue]
+  # end
+  # ```
+  module CacheKeyLoader
+    # Load a single keys for a cache fetcher
+    #
+    # @param cache_fetcher [_CacheFetcher]
+    # @param db_key Reference to what to load from the database.
+    # @return The database value corresponding to the database key.
+    def self.load(cache_fetcher, db_key)
+      cache_key = cache_fetcher.cache_key(db_key)
+
+      db_value = nil
+
+      cache_value = IdentityCache.fetch(cache_key) do
+        db_value = cache_fetcher.load_one_from_db(db_key)
+        cache_fetcher.cache_encode(db_value)
+      end
+
+      db_value || cache_fetcher.cache_decode(cache_value)
+    end
+
+    # Load multiple keys for a cache fetcher
+    # @param cache_fetcher [_CacheFetcher]
+    # @param db_key [Array] Reference to what to load from the database.
+    # @return [Hash] A hash mapping each database key to its corresponding value
+    def self.load_multi(cache_fetcher, db_keys)
+      cache_key_to_db_key_hash = {}
+
+      db_keys.each do |db_key|
+        cache_key = cache_fetcher.cache_key(db_key)
+        cache_key_to_db_key_hash[cache_key] = db_key
+      end
+
+      load_result = nil
+
+      cache_keys = cache_key_to_db_key_hash.keys
+      cache_result = cache_fetch_multi(cache_keys) do |unresolved_cache_keys|
+        resolve_miss_result = {}
+        load_result = resolve_multi_on_miss(cache_fetcher, unresolved_cache_keys,
+          cache_key_to_db_key_hash, resolve_miss_result)
+        resolve_miss_result
+      end
+
+      load_result ||= {}
+      cache_result.each do |cache_key, cache_value|
+        db_key = cache_key_to_db_key_hash.fetch(cache_key)
+        load_result[db_key] ||= cache_fetcher.cache_decode(cache_value)
+      end
+
+      load_result
+    end
+
+    class << self
```

Comment:
Should we put ☝️ in this `class << self` block or is there a reason why we need to use `self.method` syntax for them?

Reply:
yeah, we can put the public singleton methods in the `class << self` block and it won't be any different.